# <font color=blue>Assignments for "Understanding The Relationship"</font>

To close out this lesson, you're going to do three assignments. For the first assignment, you'll write up a short answer to a question in a Gist file.  For the second two assignments, you'll do your work in Jupyter notebooks, and you should link to those notebooks in the same Gist file.

Please submit a single Gist file containing the answer to first assignment, plus links for second two.

## 1. Interpretion and signficance

Suppose that we would like to know how much families in the US are spending on recreation annually. We estimated the following model:

$$ expenditure = 873 + 0.0012annual\_income + 0.00002annual\_income^2 - 223.57have\_kids $$

*expenditure* is the annual spending on recreation in US dollars, *annual_income* is the annual income in US dollars and *have_kids* is a dummy variable indicating the families with children. Interpret the estimated coefficients. What additional statistics should be given in order to make sure that your interpretations make sense statistically. Write up your answer and save in a Gist. 

- We can apply t-test to understand whether a coefficient is statistically significant.
- We also need to know P-values and t-values.

## 2. Weather model

In this exercise, you'll work with the historical temperature data from the previous lesson. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [2]:
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

- Build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. Are the estimated coefficients statistically significant? Are the signs of the estimated coefficients in line with your previous expectations? Interpret the estimated coefficients. What are the relations between the target and the explanatory variables? 

In [3]:
y = weather_df['Temperature (C)'] - weather_df['Apparent Temperature (C)']

X = weather_df[['Humidity', 'Wind Speed (km/h)']]

In [4]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sat, 08 May 2021   Prob (F-statistic):               0.00
Time:                        19:16:20   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

P-values for all coefficients are 0, which means that they are all statistically significant. While both coefficients have a positive sign, coefficient for humidity is larger than wind speed, which is in accordance with my expectation.

- Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS. Are the coefficients statistically significant? Did the signs of the estimated coefficients for *humidity* and *windspeed* change? Interpret the estimated coefficients.

In [5]:
X['humidity*windspeed'] = X['Humidity'] * X['Wind Speed (km/h)']

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sat, 08 May 2021   Prob (F-statistic):               0.00
Time:                        19:16:20   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity              -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)     -0.0905      0.002    -36.797      0.000      -0.095      -0.086
humidity*windspeed     0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

P-values (expect that of the constant) are still 0, which means that they are still statistically significant. However, signs of both Humidity and Wind Speed have changed, they are now negative and have smaller absolute values than earlier. The interaction variable, on the other hand, has a positive coefficient.

##  3. House prices model

In this exercise, you'll interpret your house prices model. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [6]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

- Run your house prices model again and interpret the results. Which features are statistically significant and which are not?

In [7]:
house_prices_df_null_ratios = house_prices_df.isna().sum() / len(house_prices_df)
house_prices_df.drop(house_prices_df_null_ratios[house_prices_df_null_ratios > .1].index, axis=1, inplace=True)
house_prices_df.dropna(inplace=True)

In [8]:
X = house_prices_df.select_dtypes(exclude='object').drop('SalePrice', axis=1)
y = house_prices_df['SalePrice']

In [9]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.808
Model:                            OLS   Adj. R-squared:                  0.803
Method:                 Least Squares   F-statistic:                     161.5
Date:                Sat, 08 May 2021   Prob (F-statistic):               0.00
Time:                        19:16:21   Log-Likelihood:                -15881.
No. Observations:                1338   AIC:                         3.183e+04
Df Residuals:                    1303   BIC:                         3.201e+04
Df Model:                          34                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          4.344e+05   1.49e+06      0.292      0.770   -2.48e+06    3.35e+06
Id               -0.4159      2.304     -0.180      0.857      -4.937       4.105
MSSubClass     -164.9821     28.463     -5.796      0.000    -220.820    -109.144
LotArea           0.3950      0.102      3.860      0.000       0.194       0.596
OverallQual    1.854e+04   1277.521     14.509      0.000     1.6e+04     2.1e+04
OverallCond    5539.4909   1144.985      4.838      0.000    3293.275    7785.707
YearBuilt       343.8423     78.784      4.364      0.000     189.286     498.399
YearRemodAdd    149.4128     75.737      1.973      0.049       0.833     297.992
MasVnrArea       27.9086      6.117      4.562      0.000      15.908      39.910
BsmtFinSF1       10.6647      2.743      3.888      0.000       5.284      16.046
BsmtFinSF2        1.9829      4.753      0.417      0.677      -7.342      11.308
BsmtUnfSF         0.6779      2.754      0.246      0.806      -4.725       6.081
TotalBsmtSF      13.3256      4.722      2.822      0.005       4.062      22.590
1stFlrSF         11.0700      7.818      1.416      0.157      -4.268      26.408
2ndFlrSF         18.5493      6.803      2.727      0.006       5.203      31.896
LowQualFinSF      0.9313     18.561      0.050      0.960     -35.482      37.345
GrLivArea        30.5506      6.929      4.409      0.000      16.957      44.144
BsmtFullBath   8486.5250   2773.439      3.060      0.002    3045.630    1.39e+04
BsmtHalfBath   2007.5870   4271.709      0.470      0.638   -6372.593    1.04e+04
FullBath       1106.9972   3071.372      0.360      0.719   -4918.378    7132.373
HalfBath      -3940.6439   2848.081     -1.384      0.167   -9527.970    1646.682
BedroomAbvGr  -9641.7992   1875.348     -5.141      0.000   -1.33e+04   -5962.766
KitchenAbvGr  -2.589e+04   6633.844     -3.903      0.000   -3.89e+04   -1.29e+04
TotRmsAbvGrd   5615.9799   1302.811      4.311      0.000    3060.144    8171.816
Fireplaces     3777.7691   1852.760      2.039      0.042     143.051    7412.487
GarageYrBlt     -48.8651     81.702     -0.598      0.550    -209.147     111.417
GarageCars     1.524e+04   3079.095      4.948      0.000    9195.805    2.13e+04
GarageArea        4.8423     10.400      0.466      0.642     -15.560      25.245
WoodDeckSF       23.7624      8.253      2.879      0.004       7.572      39.952
OpenPorchSF      -6.8007     16.325     -0.417      0.677     -38.827      25.226
EnclosedPorch    11.1010     17.862      0.621      0.534     -23.941      46.143
3SsnPorch        23.2182     32.087      0.724      0.469     -39.730      86.166
ScreenPorch      54.7314     17.391      3.147      0.002      20.614      88.849
PoolArea        -36.8969     23.959     -1.540      0.124     -83.899      10.105
MiscVal          -0.6979      1.909     -0.366   

- Now, exclude the insignificant features from your model. Did anything change?

In [10]:
X_significant = X[['MSSubClass', 'LotArea', 'OverallQual', 'OverallCond',
                   'YearBuilt', 'MasVnrArea', 'BsmtFinSF1', 'TotalBsmtSF',
                    '2ndFlrSF', 'GrLivArea', 'BsmtFullBath', 'BedroomAbvGr',
                    'KitchenAbvGr', 'TotRmsAbvGrd', 'GarageCars', 'WoodDeckSF',
                    'ScreenPorch']]

In [11]:
results = sm.OLS(y, X_significant).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              SalePrice   R-squared (uncentered):                   0.969
Model:                            OLS   Adj. R-squared (uncentered):              0.969
Method:                 Least Squares   F-statistic:                              2450.
Date:                Sat, 08 May 2021   Prob (F-statistic):                        0.00
Time:                        19:16:21   Log-Likelihood:                         -15919.
No. Observations:                1338   AIC:                                  3.187e+04
Df Residuals:                    1321   BIC:                                  3.196e+04
Df Model:                          17                                                  
Covariance Type:            nonrobust                                                  
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
MSSubClass    -148.4595     28.533     -5.203      0.000    -204.434     -92.485
LotArea          0.3647      0.102      3.565      0.000       0.164       0.565
OverallQual   2.256e+04   1187.777     18.990      0.000    2.02e+04    2.49e+04
OverallCond   3176.0287    951.836      3.337      0.001    1308.754    5043.304
YearBuilt      -35.3897      5.659     -6.254      0.000     -46.491     -24.289
MasVnrArea      30.0674      6.088      4.939      0.000      18.124      42.011
BsmtFinSF1      10.7016      3.186      3.359      0.001       4.452      16.952
TotalBsmtSF     21.6900      5.914      3.667      0.000      10.087      33.293
2ndFlrSF        10.6670      6.030      1.769      0.077      -1.163      22.497
GrLivArea       32.9011      6.386      5.152      0.000      20.373      45.429
BsmtFullBath  9176.0629   2530.796      3.626      0.000    4211.246    1.41e+04
BedroomAbvGr -9779.5379   1838.564     -5.319      0.000   -1.34e+04   -6172.713
KitchenAbvGr -3.843e+04   6432.066     -5.975      0.000   -5.11e+04   -2.58e+04
TotRmsAbvGrd  5541.8748   1316.871      4.208      0.000    2958.489    8125.261
GarageCars    2.139e+04   2061.265     10.378      0.000    1.73e+04    2.54e+04
WoodDeckSF      31.4918      8.208      3.836      0.000      15.389      47.595
ScreenPorch     41.3815     17.170      2.410      0.016       7.697      75.066
==============================================================================
Omnibus:                      726.855   Durbin-Watson:                   1.961
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           110445.710
Skew:                          -1.466   Prob(JB):                         0.00
Kurtosis:                      47.413   Cond. No.                     9.92e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.92e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R-squared has improved significantly.

- Interpret the statistically significant coefficients by quantifying their relations with the house prices. Which features have more prominent effect on the house prices?

While some coefficients did not change much, some others have changed a lot. For instance, coefficient of the variable "OverallQual," which has the highest effect on the target variable, has improved from 18540 to 22560. "OverallCond," on the other hand, decreased from 5539.4909 to 3176.0287. Two other variables with highest coefficients, hence the most prominent effects after "OverallQual" are "BsmtFullBath" and "TotRmsAbvGrd."

- Do the results sound reasonable to you? If not, try to explain the potential reasons.

Yes, I think they are reasonable.